src/id.rs

```diff
@@ -38,23 +39,79 @@ impl Id {
     #[doc(hidden)]
     #[track_caller]
     #[inline]
-    pub const unsafe fn from_u32(v: u32) -> Self {
+    pub const unsafe fn from_data(v: u32) -> Self {
         debug_assert!(v < Self::MAX_U32);
+
         Id {
-            // SAFETY: Caller obligation
-            value: unsafe { NonZeroU32::new_unchecked(v + 1) },
+            // SAFETY: Caller obligation.
+            value: unsafe { NonZeroU64::new_unchecked((v + 1) as u64) },
         }
     }
 
+    /// Create a `salsa::Id` from a u64 value.
+    ///
+    /// This should only be used to recreate an `Id` together with `Id::as_u64`.
+    ///
+    /// # Safety
+    ///
+    /// The data bits of the supplied value must represent a valid `Id` returned
+    /// by `Id::as_u64`.
+    #[doc(hidden)]
+    #[track_caller]
     #[inline]
-    pub const fn as_u32(self) -> u32 {
-        self.value.get() - 1
+    pub const unsafe fn from_bits(v: u64) -> Self {
+        Id {
+            // SAFETY: Caller obligation.
+            value: unsafe { NonZeroU64::new_unchecked(v) },
+        }
+    }
+
+    /// Mark the `Id` with a generation.
+    ///
+    /// This `Id` will refer to the same page and slot in the database,
+    /// but will differ from other identifiers of the slot based on the
+    /// provided generation.
+    #[inline]
+    pub fn with_generation(self, generation: u32) -> Id {
+        let mut value = self.value.get();
+
+        value &= 0xFFFFFFFF;
+        value |= (generation as u64) << 32;
+
+        Id {
+            // SAFETY: The niche of `value` is in the lower bits, which we did not touch.
+            value: unsafe { NonZeroU64::new_unchecked(value) },
+        }
+    }
+
+    /// Return the data portion of this `Id`.
+    #[inline]
+    pub const fn data(self) -> u32 {
```

Comment:
Bikeshedding: Maybe rename this to `index`. An id consists of a generation and its index into the database. This seems clearer to me than the generic data. The `index` terminology would also match with what you use in `DatabaseKeyIndex`

Reply:
Done.